interoperator/pkg/watches/channel.go

```diff
@@ -0,0 +1,252 @@
+	"k8s.io/client-go/tools/clientcmd"
+	kubernetes "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+// createWatchChannel creates a channel and sets watches on all the
+// clusters for all resources. Return the channel on which events are sent
+// and a channel for stopping the watches
+func createWatchChannel(controllerName string, defaultCluster kubernetes.Client, clusterRegistry registry.ClusterRegistry,
+	resources []osbv1alpha1.APIVersionKind) (<-chan event.GenericEvent, chan<- struct{}, error) {
+
+	events := make(chan event.GenericEvent, 1024)
+	stopCh := make(chan struct{})
+	var waitgroup sync.WaitGroup
+
+	gvkList := make([]schema.GroupVersionKind, len(resources))
+	for i, resource := range resources {
+		gv, err := schema.ParseGroupVersion(resource.APIVersion)
+		if err != nil {
+			log.Error(err, "unable to parse APIVersion. ignore watching", "APIVersion",
+				resource.APIVersion, "controller", controllerName)
+			continue
+		}
+		gvk := gv.WithKind(resource.Kind)
+		gvkList[i] = gvk
+	}
+
+	clusterList, err := clusterRegistry.ListClusters(nil)
+	if err != nil {
+		log.Error(err, "unable to get list clusters", "controller", controllerName)
+		return nil, nil, err
+	}
+	clusters := clusterList.Items
+	clusterWatcherList := make([]*clusterWatcher, len(clusters))
+	for i, cluster := range clusters {
+		configBytes, err := cluster.GetKubeConfig(defaultCluster)
+		if err != nil {
+			log.Error(err, "unable to fetch kubeconfig. Not watching on cluster.",
+				"clusterID", cluster.GetName(), "namespace", cluster.GetNamespace(), "controller", controllerName)
+			continue
+		}
+		cfg, err := clientcmd.RESTConfigFromKubeConfig(configBytes)
+		if err != nil {
+			log.Error(err, "unable to decode kubeconfig. Not watching on cluster.",
+				"clusterID", cluster.GetName(), "controller", controllerName)
+			continue
+		}
+
+		dynamicClient, err := dynamic.NewForConfig(cfg)
+		if err != nil {
+			log.Error(err, "unable to create dynamic client. Not watching on cluster.",
+				"clusterID", cluster.GetName(), "controller", controllerName)
+			continue
+		}
+
+		discoveryClient, err := discovery.NewDiscoveryClientForConfig(cfg)
+		if err != nil {
+			log.Error(err, "unable to create discovery client. Not watching on cluster.",
+				"clusterID", cluster.GetName(), "controller", controllerName)
+			continue
+		}
+
+		stop := make(chan struct{})
+		clusterWatcherList[i] = &clusterWatcher{
+			controllerName:  controllerName,
+			clusterID:       cluster.GetName(),
+			dynamicClient:   dynamicClient,
+			discoveryClient: discoveryClient,
+			gvkList:         gvkList,
+			events:          events,
+			stop:            stop,
+			stopper:         stop,
+			waitgroup:       &waitgroup,
+		}
+		err = clusterWatcherList[i].start()
+		if err != nil {
+			log.Error(err, "failed to start watch on cluster",
+				"clusterID", cluster.GetName(), "controller", controllerName)
+			continue
+		}
+	}
+	go func() {
+		<-stopCh
```

Comment:
Why not simply pass the `stopCh` [above](https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/673/files#diff-4e65bc02c98373a2de32876b47339db5R87) avoid this goroutine?

Reply:
The stop is signalled by a write to the channel (Not closing the channel). So a event sent to stopCh must be forwarded to
many go routines (one routine per cluster per gvk). If we pass the channel only one routine will read the stop event. For multiplexing the stop event this is required.

Reply:
> If we pass the channel to all go routines, only one routine will read the stop event. For multiplexing the stop event this is required.

Why not close the channel?

Reply:
Will update